src/stdlib/mod.rs

```diff
@@ -81,12 +55,58 @@ pub const QUALIFIED_BUILTINS: &str = "std/builtins";
 const ASSERT_FN: &str = "assert(condition: Bool)";
 const ASSERT_EQ_FN: &str = "assert_eq(lhs: Field, rhs: Field)";
 
-pub const BUILTIN_FNS_DEFS: [(&str, FnHandle); 2] =
-    [(ASSERT_EQ_FN, assert_eq), (ASSERT_FN, assert)];
+pub static BUILTIN_FNS_SIGS: Lazy<HashMap<&'static str, FnSig>> = Lazy::new(|| {
+    let sigs = [ASSERT_FN, ASSERT_EQ_FN];
+
+    // create a hashmap from the FnSig
+    sigs.iter()
+        .map(|sig| {
+            let ctx = &mut ParserCtx::default();
+            let mut tokens = Token::parse(0, sig).unwrap();
+            let fn_sig = FnSig::parse(ctx, &mut tokens).unwrap();
+
+            (sig.to_owned(), fn_sig)
+        })
+        .collect()
+});
+
+pub fn get_builtin_fn<B>(name: &str) -> FnInfo<B>
+where
+    B: Backend,
+{
+    let ctx = &mut ParserCtx::default();
+    let mut tokens = Token::parse(0, name).unwrap();
+    let sig = FnSig::parse(ctx, &mut tokens).unwrap();
+
+    let fn_handle = match name {
+        ASSERT_FN => assert,
+        ASSERT_EQ_FN => assert_eq,
+        _ => unreachable!(),
+    };
+
+    FnInfo {
+        kind: FnKind::BuiltIn(sig, fn_handle),
+        span: Span::default(),
+    }
+}
+
+/// a function iterate through builtin functions
+pub fn builtin_fns<B: Backend>() -> Vec<FnInfo<B>> {
+    BUILTIN_FNS_SIGS
+        .iter()
+        .map(|(sig, _)| get_builtin_fn::<B>(sig))
+        .collect()
+}
+
+pub fn has_builtin_fn(name: &str) -> bool {
+    BUILTIN_FNS_SIGS.iter().any(|(_, s)| s.name.value == name)
+}
```

Comment:
can you create a PR that we could merge first that refactors the builtins? This would simplify this PR :o